### PR TITLE
HPCC-25473 Common up Copy and Download Selection

### PR DESCRIPTION
**Pull request by GordonSmith**
Signed-off-by: Gordon Smith <GordonJSmith@gmail.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@GordonSmith Needs rebasing

**Comment by GordonSmith** [comment]
@richardkchapman rebased

**Review by jeclrsg — DISMISSED** [review]
@GordonSmith Looks good to me. Couple of questions about whether a few more lines could have been removed. I wasn't sure though.

File: esp/src/src-react/components/InfoGrid.tsx [hunk 1]
```diff
@@ -48,9 +49,9 @@ export const InfoGrid: React.FunctionComponent<InfoGridProps> = ({
     const [otherChecked, setOtherChecked] = React.useState(true);
     const [filterCounts, setFilterCounts] = React.useState<any>({});
     const [grid, setGrid] = React.useState<any>(undefined);
-    const [visibleExceptions, setvisibleExceptions] = React.useState<WUInfo.ECLException[]>([]);
+    const [, setvisibleExceptions] = React.useState<WUInfo.ECLException[]>([]);
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply esp/src/src-react/components/InfoGrid.tsx:52]: Probably - will revisit...

File: esp/src/src-react/components/Variables.tsx [hunk 2]
```diff
@@ -24,7 +23,7 @@ export const Variables: React.FunctionComponent<VariablesProps> = ({
 
     const [grid, setGrid] = React.useState<any>(undefined);
     const [selection, setSelection] = React.useState([]);
-    const [uiState, setUIState] = React.useState({ ...defaultUIState });
+    const [, setUIState] = React.useState({ ...defaultUIState });
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply esp/src/src-react/components/Variables.tsx:26]: Again probably - the uiState may come into play if we ever added more actions - but suspect not...  will revisit...

**Comment by GordonSmith** [comment]
Changes made as requested.

File: esp/src/src-react/components/Common.tsx [hunk 3]
```diff
@@ -1,4 +1,24 @@
 import * as React from "react";
 import { VerticalDivider } from "@fluentui/react";
+import * as Utility from "src/Utility";
+import nlsHPCC from "src/nlsHPCC";
 
 export const ShortVerticalDivider = () => <VerticalDivider styles={{ divider: { paddingTop: "20%", height: "60%" } }} />;
+
+export function createCopyDownloadSelection(grid, selection: any, filename: string) {
```
Comment (jbrundage) [inline esp/src/src-react/components/Common.tsx:8]: What is `grid`? It should probably be typed.

File: esp/src/src/Utility.ts [hunk 4]
```diff
@@ -209,6 +210,38 @@ export function espSkew2NumberTests() {
     }, this);
 }
 
+export function formatAsDelim(grid, rows: any, delim = ",") {
```
Comment (jbrundage) [inline esp/src/src/Utility.ts:213]: Couldn't `hpccjs.common.csvFormat` be used here?

File: esp/src/src-react/components/Common.tsx (hunk 3 above)
Comment (GordonSmith) [reply esp/src/src-react/components/Common.tsx:8]: grid is a DGrid and doesn't have a type.

File: esp/src/src/Utility.ts (hunk 4 above)
Comment (GordonSmith) [reply esp/src/src/Utility.ts:213]: I don't see a csvFormat in common?

**Review by jeclrsg — APPROVED** [review]

**Pull request merged**